cluster-autoscaler/cloudprovider/mcm/mcm_manager.go

```diff
@@ -404,174 +409,231 @@ func (m *McmManager) GetMachineDeploymentSize(machinedeployment *MachineDeployme
-			} else {
-				expectedToTerminateMachineNodePairs = append(expectedToTerminateMachineNodePairs, machineNameNodeNamePair{mclone.Name, mclone.Labels["node"]})
+			val, ok := machine.Annotations[machinePriorityAnnotation]
+			if ok && val != "3" {
+				ctx, cancelFn := context.WithDeadline(context.Background(), time.Now().Add(m.maxRetryTimeout))
+				defer cancelFn()
+				for {
+					mc, err := m.machineLister.Machines(m.namespace).Get(machine.Name)
+					if err != nil {
+						klog.Errorf("Unable to fetch Machine object %s, Error: %+v", mc.Name, err)
+						return err
+					}
+					clone := mc.DeepCopy()
+					clone.Annotations[machinePriorityAnnotation] = "3"
+					_, err = m.machineClient.Machines(m.namespace).Update(ctx, clone, metav1.UpdateOptions{})
+					if err != nil {
+						select {
+						case <-ctx.Done():
+							return fmt.Errorf("could not reset priority annotation on machine %s; timed out, error: %+v", clone.Name, err)
+						case <-time.After(m.conflictRetryInterval):
+							klog.Warningf("could not reset priority annotation for machine %s; Error: %+v, will retry the operation", clone.Name, err)
+							continue
+						}
+					}
+					break
+				}
 			}
+			return nil
+		}()
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
 
-			if mclone.Annotations != nil {
-				if mclone.Annotations[machinePriorityAnnotation] == "1" {
-					klog.Infof("Machine %q priority is already set to 1, hence skipping the update", machine.Name)
+// prioritizeMachinesForDeletion prioritizes the targeted machines by updating their priority annotation to 1
+func (m *McmManager) prioritizeMachinesForDeletion(targetMachineRefs []*Ref, mdName string) ([]machineNameNodeNamePair, error) {
+	var expectedToTerminateMachineNodePairs []machineNameNodeNamePair
+	for _, machineRef := range targetMachineRefs {
+		// Trying to update the priority of machineRef till m.maxRetryTimeout
+		ctx, cancelFn := context.WithDeadline(context.Background(), time.Now().Add(m.maxRetryTimeout))
+		err := func() error {
+			defer cancelFn()
+			for {
+				mc, err := m.machineLister.Machines(m.namespace).Get(machineRef.Name)
+				if err != nil {
+					klog.Errorf("Unable to fetch Machine object %s, Error: %+v", machineRef.Name, err)
+					return err
+				}
+				if isMachineTerminating(mc) {
 					break
 				}
-				mclone.Annotations[machinePriorityAnnotation] = "1"
-			} else {
-				mclone.Annotations = make(map[string]string)
-				mclone.Annotations[machinePriorityAnnotation] = "1"
+				expectedToTerminateMachineNodePairs = append(expectedToTerminateMachineNodePairs, machineNameNodeNamePair{mc.Name, mc.Labels["node"]})
+				if err = m.updateAnnotationOnMachineBeforeDeadline(ctx, mc.Name, machinePriorityAnnotation, "1"); err != nil {
+					return err
+				}
+				// Break out of loop when update succeeds
+				klog.Infof("Machine %s of machineDeployment %s marked with priority 1 successfully", machineRef.Name, mdName)
+				break
 			}
+			return nil
+		}()
+		if err != nil {
```

Comment:
If one of the machines could not be updated then why return an error? Either you continue or you rollback, here you decide to abruptly exit mid-way.

Reply:
Before prioritizing any machine, we reset the priority of the remaining machines. This will ensure that only the machines that need to be removed are prioritized.